xds/src/main/java/io/grpc/xds/FaultFilter.java

```diff
@@ -0,0 +1,425 @@
+import javax.annotation.Nullable;
+
+/** HttpFault filter implementation. */
+final class FaultFilter implements Filter, ClientInterceptorBuilder {
+
+  static final FaultFilter INSTANCE =
+      new FaultFilter(ThreadSafeRandomImpl.instance, new AtomicLong());
+  @VisibleForTesting
+  static final Metadata.Key<String> HEADER_DELAY_KEY =
+      Metadata.Key.of("x-envoy-fault-delay-request", Metadata.ASCII_STRING_MARSHALLER);
+  @VisibleForTesting
+  static final Metadata.Key<String> HEADER_DELAY_PERCENTAGE_KEY =
+      Metadata.Key.of("x-envoy-fault-delay-request-percentage", Metadata.ASCII_STRING_MARSHALLER);
+  @VisibleForTesting
+  static final Metadata.Key<String> HEADER_ABORT_HTTP_STATUS_KEY =
+      Metadata.Key.of("x-envoy-fault-abort-request", Metadata.ASCII_STRING_MARSHALLER);
+  @VisibleForTesting
+  static final Metadata.Key<String> HEADER_ABORT_GRPC_STATUS_KEY =
+      Metadata.Key.of("x-envoy-fault-abort-grpc-request", Metadata.ASCII_STRING_MARSHALLER);
+  @VisibleForTesting
+  static final Metadata.Key<String> HEADER_ABORT_PERCENTAGE_KEY =
+      Metadata.Key.of("x-envoy-fault-abort-request-percentage", Metadata.ASCII_STRING_MARSHALLER);
+  static final String TYPE_URL =
+      "type.googleapis.com/envoy.extensions.filters.http.fault.v3.HTTPFault";
+
+  private final ThreadSafeRandom random;
+  private final AtomicLong activeFaultCounter;
+
+  @VisibleForTesting
+  FaultFilter(ThreadSafeRandom random, AtomicLong activeFaultCounter) {
+    this.random = random;
+    this.activeFaultCounter = activeFaultCounter;
+  }
+
+  @Override
+  public String[] typeUrls() {
+    return new String[] { TYPE_URL };
+  }
+
+  @Override
+  public StructOrError<FaultConfig> parseFilterConfig(Message rawProtoMessage) {
+    HTTPFault httpFaultProto;
+    if (!(rawProtoMessage instanceof Any)) {
+      return StructOrError.fromError("Invalid config type: " + rawProtoMessage.getClass());
+    }
+    Any anyMessage = (Any) rawProtoMessage;
+    try {
+      httpFaultProto = anyMessage.unpack(HTTPFault.class);
+    } catch (InvalidProtocolBufferException e) {
+      return StructOrError.fromError("Invalid proto: " + e);
+    }
+    return parseHttpFault(httpFaultProto);
+  }
+
+  private static StructOrError<FaultConfig> parseHttpFault(HTTPFault httpFault) {
+    FaultDelay faultDelay = null;
+    FaultAbort faultAbort = null;
+    if (httpFault.hasDelay()) {
+      faultDelay = parseFaultDelay(httpFault.getDelay());
+    }
+    if (httpFault.hasAbort()) {
+      StructOrError<FaultAbort> faultAbortOrError = parseFaultAbort(httpFault.getAbort());
+      if (faultAbortOrError.errorDetail != null) {
+        return StructOrError.fromError(
+            "HttpFault contains invalid FaultAbort: " + faultAbortOrError.errorDetail);
+      }
+      faultAbort = faultAbortOrError.struct;
+    }
+    String upstreamCluster = httpFault.getUpstreamCluster();
+    List<String> downstreamNodes = httpFault.getDownstreamNodesList();
```

Comment:
Do we need upstream cluster and downstream nodes?

Reply:
Not needed now. Removed.